Try to Compile LAMMPS with OBMD plugin in github action

laraPPr opened the pull request.

laraPPr committed:
Create MultiXscale:compile_lammps

laraPPr committed:
Update MultiXscale:compile_lammps

laraPPr committed:
Update MultiXscale:compile_lammps

laraPPr committed:
Update MultiXscale:compile_lammps

laraPPr committed:
Create MultiXscale:compile_lammps.yml

laraPPr committed:
Delete .github/workflows/MultiXscale:compile_lammps

laraPPr committed:
Add lammps config file

laraPPr committed:
Update LAMMPS_OBMD-foss-2023a-software-commit.eb

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,26 @@
+        module avail LAMMPS
+        module load LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        ls /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        lmp -in /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD/examples/OBMD/in.OBMD
```
I noticed this is only using 1 CPU, but actually there are typically 4 available. Is there a way to enable MPI or OpenMP and tell it to use whatever is available?

ocaisa replied:
```suggestion
        mpiexec -n 4 lmp -in /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD/examples/OBMD/in.OBMD
```
see https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,26 @@
+      - 'src/OBMD/**'
+      - 'examples/OBMD/**'
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: eessi/github-action-eessi@v3
+    - name: set up EasyBuild
+      run: |
+        module load EasyBuild/4.9.4
+        module load EESSI-extend/2023.06-easybuild
+        export EASYBUILD_TRACE=1
+        eb --show-config
+        echo "Get EasyConfig"
+        wget https://raw.githubusercontent.com/laraPPr/lammps/refs/heads/OBMD_compile_in_CI/.github/workflows/LAMMPS_OBMD-foss-2023a-software-commit.eb
```
```suggestion
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os:
          - ubuntu-24.04-arm
          - ubuntu-24.04
    steps:
    - uses: actions/checkout@v4
    - uses: eessi/github-action-eessi@v3
    - name: set up EasyBuild
      run: |
        module load EESSI-extend/2023.06-easybuild
        eb --show-config
```

ocaisa replied:
Loading the latest EasyBuild is done by default and `--trace` is also already set.

You needed to download the easyconfig because you didn't actually check out the repository. Having said that, I wonder if it is not a better idea to pull in the development easyconfig from the dev.eessi.io repo so it doesn't have to be updated in two places?

laraPPr replied:
The two configs are a little different from each other with one I use the patch and with the other I just use the sources in the branch of the pr. But I was thinking about wanting them to remain the same or having two config managed in a the dev.eessi.io-LAMMPS-plugins (still don't really have a name for it). One that uses a patch and the other that uses the sources.

ocaisa replied:
Yeah, you were missing
```
    - uses: actions/checkout@v4
```
in your workflow, so you never actually check out the repo (I included it  in my suggestion above).

ocaisa replied:
To keep them the same, you can just have a script that generates the patch and run that before the `eb` command? You'd need something like that anyway when it comes to `dev.eessi.io`

laraPPr replied:
Ok yes won't be able to set that up today anymore but I will look into setting that up. I think I already a something if I remember correctly.

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,26 @@
+        eb --show-config
+        echo "Get EasyConfig"
+        wget https://raw.githubusercontent.com/laraPPr/lammps/refs/heads/OBMD_compile_in_CI/.github/workflows/LAMMPS_OBMD-foss-2023a-software-commit.eb
+        eb LAMMPS_OBMD-foss-2023a-software-commit.eb --software-commit ${{ github.sha }} --include-easyblocks-from-commit 3d295a926318468e3e8fdd99fdcfea02cd796403
```
Depending on what you decide to do, you may need to change this to
```
eb .github/workflows/LAMMPS_OBMD-foss-2023a-software-commit.eb ...
```

laraPPr replied:
I tried that and it did not work. For some reason that file did not exist on the vm where action runs.

laraPPr replied:
See https://github.com/laraPPr/lammps/actions/runs/12989912297/job/36224164791

ocaisa replied:
Yeah, you were missing
```
    - uses: actions/checkout@v4
```
in your workflow, so you never actually check out the repo (I included it  in my suggestion above).

laraPPr committed:
Update .github/workflows/MultiXscale:compile_lammps.yml

Co-authored-by: ocaisa <alan.ocais@cecam.org>

laraPPr committed:
Update MultiXscale:compile_lammps.yml

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,26 @@
+        module avail LAMMPS
+        module load LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        ls /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        mpirun -np 4 lmp -in /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD/examples/OBMD/in.OBMD
```
I suspect there are not really 4 cores, 4 virtual but only 2 physical. You can either
```
mpirun -np 4 --use-hwthread-cpus
```
or 
```
mpirun -np 4 --oversubscribe
```
or
```
mpirun -np 2
```

laraPPr committed:
source repo and oversubscribe

laraPPr commented:
`examples/balance/in.balance` fails on arm not sure why. I haven't seen this before and I probably can't get to the logs and in this action

laraPPr committed:
make the easybuild log more verbose in stdout

laraPPr committed:
use artifacts

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
try to use download artifacts

laraPPr committed:
add timestep patch to banlance example

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,31 @@
+        eb .github/workflows/LAMMPS_OBMD-foss-2023a-software-commit.eb --software-commit ${{ github.sha }} --include-easyblocks-from-commit 3d295a926318468e3e8fdd99fdcfea02cd796403
+        module avail LAMMPS
+        module load LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        ls /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
```
Maybe just search for `$HOME` in `$EBROOTLAMMPS`?
```
echo $EBROOTLAMMPS | grep $HOME
```

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,31 @@
+        module avail LAMMPS
+        module load LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        ls /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        mpirun -np 4 --oversubscribe lmp -in /home/runner/eessi/versions/2023.06/software/linux/x86_64/amd/zen3/software/LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD/examples/OBMD/in.OBMD
```
Same here, use `$EBROOTLAMMPS`

laraPPr replied:
ah yes thats indeed better will fix

laraPPr committed:
use $EESSI_SOFTWARE_SUBDIR

laraPPr committed:
use $EBROOTLAMMPS

laraPPr committed:
remove downloading artefacts because is skipped when the run fails

Wanted to see if it was possible to get the easibuild logs when it fail but artefacts does not seem to be the tool to use in this case.

laraPPr commented:
Wow had not expected that the arm instance would almost be 10 min faster than the x86-64

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,29 @@
+        eb .github/workflows/LAMMPS_OBMD-foss-2023a-software-commit.eb --software-commit ${{ github.sha }} --include-easyblocks-from-commit 3d295a926318468e3e8fdd99fdcfea02cd796403
+        module avail LAMMPS
+        module load LAMMPS/${{ github.sha }}-foss-2023a-kokkos-dev_OBMD
+        ls $EBROOTLAMMPS
```
```suggestion
        echo $EBROOTLAMMPS | grep $HOME
        ls $EBROOTLAMMPS
```

ocaisa replied:
This is not actually _checking_ that the installation comes from where you think, that's what I suggested the grep

laraPPr committed:
Update .github/workflows/MultiXscale:compile_lammps.yml

Co-authored-by: ocaisa <alan.ocais@cecam.org>

laraPPr committed:
use new commit of easyblock

laraPPr committed:
Update MultiXscale:compile_lammps.yml

ocaisa commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,30 @@
+        module load EESSI-extend/2023.06-easybuild
+        eb --show-config
+        echo "Get EasyConfig"
+        eb .github/workflows/LAMMPS_OBMD-foss-2023a-software-commit.eb --software-commit ${{ github.sha }} --include-easyblocks-from-commit 55dc794b03f5d75a9ab31bc41b960abe1605134a --include-easyblocks lammps.py
```
Where is this commit coming from? Latest commit I see is `362b4679193612e04abe336fa041e2a34d183991`

laraPPr replied:
here https://github.com/easybuilders/easybuild-easyblocks/pull/3485/commits/55dc794b03f5d75a9ab31bc41b960abe1605134a

laraPPr replied:
Put back the second to last one. It seems like the merge with upstream EasyBlocks was to big for github actions to deal with.

laraPPr replied:
The commit that I'm using is from a pr that is not merge yet https://github.com/easybuilders/easybuild-easyblocks/pull/3485

laraPPr replied:
Ah this is the actual error
```
ERROR: Failed to import easyblock easybuild.easyblocks.generic.easyblock_specific from /tmp/eb-w0y1rq44/included-easyblocks-5iev4nma/easybuild/easyblocks/generic: no module named 'test.easyblocks'
```

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Create MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

laraPPr committed:
Update MultiXscale:generate_patch_file.yml

Commit:
add patch file workflow to compile workflow and get easyconfig from multixscale/dev.eessi.io-lammps-plugin-obmd

laraPPr committed:
copy patch file to directory where easybuild will find it

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
update path to patch

laraPPr committed:
Use commit used in dev.eessi.io-example

laraPPr committed:
Use the commit used in eessi.dev.io-example

laraPPr committed:
Use other lammps commit

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Update MultiXscale:compile_lammps.yml

laraPPr committed:
Use LAMMPS_COMMIT set variable in actions

laraPPr committed:
use two different variable for LAMMPS_COMMIT

laraPPr committed:
Check env.LAMMPS_COMMIT_...

laraPPr committed:
Find where the variable are stored

laraPPr committed:
Test environment variables

laraPPr committed:
Change module name by using variable

laraPPr commented on .github/workflows/MultiXscale:compile_lammps.yml:
```diff
@@ -0,0 +1,54 @@
+        module avail LAMMPS
+        module load LAMMPS/${{ vars.LAMMPS_COMMIT_BUILD }}-foss-2023a-kokkos-dev_OBMD
+        echo $EBROOTLAMMPS
+        ls $EBROOTLAMMPS
```
push logs to somewhere else after build when certain label is set?
